elastic/kibana
# Add install integrations endpoint

Resolves #183472
Resolves #183020

## Summary

1. Decouples integration installation from fleet privilege
2. Creates API endpoint to install all selected integrations as a single request:

  - Should accept list of selected integrations and log files
  - Should respond in a format easy to parse using native bash / standard unix commands like awk/sed/grep
  - Should return unified config for Elastic Agent for all installed integrations
  - ~Should update saved object with selected integrations and installation status~ Waiting for designs/requirements

## Screenshot

```text
curl --request POST \
    --url "http://localhost:5601/internal/observability_onboarding/flow/598bc802-0616-47c2-8895-c9cc24b959dd/integrations/install" \
    --header "Authorization: ApiKey emRMWHBvOEJOMmJEaFRKNnN4LS06SVJwcldSTkxTUjZtU1VpNXRLU2ZBdw==" \
    --header "Content-Type: text/tab-separated-values" \
    --data $'system\tregistry\nproduct_service\tcustom\t/path/to/access.log\ncheckout_service\tcustom\t/path/to/access.log\ncheckout_service\tcustom\t/path/to/error.log'

outputs:
  default:
    type: elasticsearch
    hosts:
      - 'http://localhost:9200'
    api_key: 'zdLXpo8BN2bDhTJ6sx--:IRprWRNLSR6mSUi5tKSfAw'
inputs:
  - id: logfile-system.auth-96f640d3-2365-4008-b634-dcbe8278b583
    data_stream:
      dataset: system.auth
      type: logs
    paths:
      - /var/log/auth.log*
      - /var/log/secure*
    exclude_files:
      - .gz$
    multiline:
      pattern: ^s
      match: after
    tags:
      - system-auth
    processors:
      - add_locale: null
  - id: logfile-system.syslog-96f640d3-2365-4008-b634-dcbe8278b583
    data_stream:
      dataset: system.syslog
      type: logs
    paths:
      - /var/log/messages*
      - /var/log/syslog*
      - /var/log/system*
    exclude_files:
      - .gz$
    multiline:
      pattern: ^s
      match: after
    processors:
      - add_locale: null
  - id: custom-logs-4e07e609-ba8e-4dbe-9490-0b4aaf9e637b
    type: logfile
    data_stream:
      namespace: default
    streams:
      - id: logs-onboarding-product_service
        data_stream:
          dataset: product_service
        paths:
          - /path/to/access.log
  - id: custom-logs-c665eb58-effe-4530-be01-8b510f969140
    type: logfile
    data_stream:
      namespace: default
    streams:
      - id: logs-onboarding-checkout_service
        data_stream:
          dataset: checkout_service
        paths:
          - /path/to/access.log
          - /path/to/error.log
```

<img width="1228" alt="Screenshot 2024-05-23 at 20 05 59" src="https://github.com/elastic/kibana/assets/190132/c2759491-b9ae-4b89-8f24-e196708d76f6">

## Testing

1. Start a Quickstart onboarding flow and copy the onboarding ID from DEV tools network tab
2. Create a new API key
3. Run the following curl:

```bash
curl --request POST \
  --url "http://localhost:5601/internal/observability_onboarding/flow/${ONBOARDING_ID}/integrations/install" \
  --header "Authorization: ApiKey ${ENCODED_API_KEY}" \
  --header "Content-Type: text/tab-separated-values" \
  --data $'system\tregistry\nproduct_service\tcustom\t/path/to/access.log\ncheckout_service\tcustom\t/path/to/access.log'
```

4. Go to installed integrations page
5. You should see the the system integration and 2 custom integrations installed.
